spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala

```diff
@@ -0,0 +1,149 @@
+    (-Math.log(scales(num_levels - 1).asInstanceOf[Double])/Math.log(2.0)).toInt
+  } else {
+    throw new IllegalArgumentException("currently only Double and Float types are supported")
+  }
+
+  private def levelMapping(
+    k_min: Int,
+    k_max: Int,
+    rois: Tensor[T],
+    canonical_scale: Int = 224,
+    canonical_level: Int = 4,
+    eps: Float = 1e-6f
+    ): Array[Int] = {
+    val s0 = canonical_scale
+    val lvl0 = canonical_level
+
+    val target_lvls = new Array[Int](rois.size(1))
+    for (i <- 1 to rois.size(1)) {
+      val a = if (classTag[T] == classTag[Float]) {
+        area(rois(i)).asInstanceOf[Float].toDouble
+      } else if (classTag[T] == classTag[Double]) {
+        area(rois(i)).asInstanceOf[Double]
+      } else {
+        throw new IllegalArgumentException("currently only Double and Float types are supported")
+      }
+
+      val s = Math.sqrt(a)
+      var target_lvl = Math.floor(lvl0 + Math.log(s / s0 + eps) / Math.log(2))
+      target_lvl = Math.min(Math.max(target_lvl, k_min), k_max)
+      target_lvls(i - 1) = (target_lvl - k_min).toInt
+    }
+
+    target_lvls
+  }
+
+  private def area(roi: Tensor[T]): T = {
+    require(roi.size().length == 1 && roi.size(1) == 5,
+      "ROI bounding box should be 1 dimensional and have 5 elements")
+    val xlow = roi.valueAt(2)
+    val ylow = roi.valueAt(3)
+    val xhigh = roi.valueAt(4)
+    val yhigh = roi.valueAt(5)
+
+    val area = ev.times(ev.plus(ev.minus(xhigh, xlow), ev.fromType(1)),
+      ev.plus(ev.minus(yhigh, ylow), ev.fromType(1)))
+    area
+  }
+
+  override def updateOutput(input: Table): Tensor[T] = {
+    val feature_maps = input[Table](1)
+    val rois = input[Tensor[T]](2)
+
+    val roi_levels = levelMapping(lvl_min, lvl_max, rois)
+    val num_rois = rois.size(1)
+    val num_channels = feature_maps.get[Tensor[T]](1).get.size(2)
+
+    output.resize(num_rois, num_channels, resolution, resolution)
+      .fill(ev.fromType[Float](Float.MinValue))
+
+    for (level <- 0 until num_levels) {
+      val feature_per_level = feature_maps.get[Tensor[T]](level + 1).get
+      val rois_ind_per_level = roi_levels.zipWithIndex.filter(_._1 == level).map(_._2)
+      val num_rois_per_level = rois_ind_per_level.length
+
+      if (num_rois_per_level > 0) {
+        val rois_per_level = Tensor[T](Array(num_rois_per_level, 5))
+        for (i <- 0 until num_rois_per_level) {
+          rois_per_level(i + 1) = rois(rois_ind_per_level(i) + 1)
+        }
+
+        val res = poolers(level).forward(T(feature_per_level, rois_per_level))
+        for (i <- 0 until num_rois_per_level) {
+          output(rois_ind_per_level(i) + 1) = res(i + 1)
+        }
+      }
+    }
+
+    output
+  }
+
+  override def updateGradInput(input: Table, gradOutput: Tensor[T]): Table = {
+    gradInput = null
+    gradInput
+  }
+
+  override def toString: String = "nn.Pooler"
+
+  override def clearState(): this.type = {
+    super.clearState()
+    this
```

Comment:
clear state for each ROIAlign in pooler array

Reply:
Added